File: coding_int/chapter_2/2_3_Delete_Middle_Node.cpp
#include<iostream>
#include<vector>
#include<string>

class node{
    public:

    node* next;
    std::string data;
};

class llist{

    private:

    node* head;
    node* tail;

    public:

    llist() {
        head = NULL;
        tail = NULL;
    }

    void addValue(std::string value){

        node *temp = new node;
        temp->data = value;
        temp->next = NULL;
        
        if (head == NULL) {
            head = temp;
            tail = temp;
        } else {
            tail->next = temp;
            tail = temp;
        }

    }

    void printValues() {

        node *temp = new node;
        temp = head;

        while (temp) {
            std::cout << temp->data << "\t";
            temp = temp->next;
        }
        std::cout << std::endl;
    }

    node* extractNode_by_value(std::string datafind) {

        node *temp = head;

        while(temp) {
            if (temp->data == datafind) {
                return temp;
            } else {
                temp = temp->next;
            }
        }
    }

};

void remove_middle_node(node *temp) {

    if (temp->next == NULL) {
        std::cout << "this is not a middle node" << std::endl;
    }
    
    node *nextnode = temp->next;

    temp->data = nextnode->data;
    temp->next = nextnode->next;

}

int main(){

    llist *testing = new llist;
    testing->addValue("test1");
    testing->addValue("test2");
    testing->addValue("test3");
    testing->addValue("test4");
    testing->addValue("test5");
    testing->addValue("test6");

    testing->printValues();

    node *middle_node;

    middle_node = testing->extractNode_by_value("test3");

    // remove the middle node
    std::cout << "removing: " << middle_node->data << std::endl;
    remove_middle_node(middle_node);

    testing->printValues();
    
    return 1;
}

File: coding_int/chapter_2/2_3_Delete_Middle_Node.py
class node:
    def __init__(self, value=None):
        self.next = None
        self.data = value

class llist:

    def __init__(self, head=None):
        self.head = head
        self.tail = None

    def __iter__(self):
        current = self.head
        while current is not None:
            yield current
            current = current.next

    def __str__(self):
        print("----start----")
        for item in self:
            if item.data is not None:
                print("{0} \t".format(item.data))

        return "----END----"

    def addValue(self, value):
        new_node = node(value)
        if self.head is None:
            self.head = new_node
            self.tail = new_node
        else:
            self.tail.next = new_node
            self.tail = new_node

def extract_node(llist, node_value):

    for item in llist:
        if item.data == node_value:
            return item

def remove_node(node):

    next_node = node.next
    node.next = next_node.next
    node.data = next_node.data

if __name__ == "__main__":

    test_list = llist()
    test_list.addValue("test1")
    test_list.addValue("test2")
    test_list.addValue("test3")
    test_list.addValue("test4")
    test_list.addValue("test5")
    test_list.addValue("test6")

    test3_node = extract_node(test_list, "test3")

    print(test3_node.data)

    remove_node(test3_node)

    print(test_list)

    